### PR TITLE
feat(backend): add postgres initialization

### PR DESCRIPTION
**Pull request by Linchin**
**Description of your changes:**
Part 1 of supporting PostgreSQL:

* Adds driver for PostgreSQL
* Supports initializing PostgreSQL
* Adds integration tests for MySQL or PostgreSQL initialization
* Moves `client_manager.go` to its own package and folder for ease of testing
* Allows users to select which type of database they want to run integration test against
* Fixes a typo in the PostgreSQL deployment guide
* Makes PostgreSQL no longer load balancer

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by Linchin** [comment]
test kubeflow-pipeline-e2e-test

File: backend/test/integration/run_tests_locally.sh
Comment (zijianjoy) [inline backend/test/integration/run_tests_locally.sh:37]: What is this part used for ? `PORT_FORWARD_PID=$!`

File: backend/test/integration/run_tests_locally.sh
Comment (zijianjoy) [inline backend/test/integration/run_tests_locally.sh:52]: NIT: How about using `set -x` at the top of file to print command automatically, instead of echoing here?

File: backend/test/integration/run_tests_locally.sh
Comment (zijianjoy) [inline backend/test/integration/run_tests_locally.sh:41]: Looks like only this line is different between `if` and `else` condition, can we append the `-runPostgreSQLTests=true` to the command if this is postgresql, so we can reduce some duplication.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [inline backend/src/apiserver/client_manager/client_manager.go:53]: It might cause a breaking change to our users for moving the db config to MySQLConfig. @chensun is it okay to make this change?

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [inline backend/src/apiserver/client_manager/client_manager.go:224]: Why do we choose `pgx` for postgresql in `DBDriverName` flag? Is there an alternative which is easier for people to understand?

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [inline backend/src/apiserver/client_manager/client_manager.go:268]: We might need to define some `const` for value like `text` and `longtext not null`.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [inline backend/src/apiserver/client_manager/client_manager.go:405]: Removing your name from logging? :)

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [inline backend/src/apiserver/client_manager/client_manager.go:396]: Do we want to set a default user like mysql does?

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [inline backend/src/apiserver/client_manager/client_manager.go:303]: I think extracting the different behavior of each driver to their own module is helpful. But feel free to defer it to future PRs.

File: backend/Dockerfile.persistenceagent
Comment (Linchin) [reply backend/Dockerfile.persistenceagent:34]: Thank you for pointing this out. I was temporarily disabling them, because the generated license files apparently changed although I didn't change this part of code. I intend to double check if we need to update all the license files.

File: backend/src/apiserver/client/sql.go
Comment (Linchin) [reply backend/src/apiserver/client/sql.go:24]: I don't have a strong preference, I'm doing this solely for style purpose. Since all these parameters are of string type, it feels more succinct to declare their type together.

File: backend/src/apiserver/client/sql.go
Comment (Linchin) [reply backend/src/apiserver/client/sql.go:49]: I think it would be great to improve security, but I wonder if it is necessary, since the DB and the backend are hosted in the same cluster.

File: backend/test/integration/run_tests_locally.sh
Comment (Linchin) [reply backend/test/integration/run_tests_locally.sh:37]: Thank you for catching this! I should have added the function that cleans up the process on exit.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (chensun) [reply backend/src/apiserver/client_manager/client_manager.go:53]: Seems to be fine, as I don't think `backend/src/apiserver/config/config.json` is user-facing?

File: backend/src/apiserver/client_manager/client_manager.go
Comment (chensun) [inline backend/src/apiserver/client_manager/client_manager.go:59]: The newly added Postgres configs seems so generic and completely covered by existing configs for mysql already. Do we need to separate them?

File: backend/src/apiserver/client_manager/client_manager.go
Comment (chensun) [inline backend/src/apiserver/client_manager/client_manager.go:15]: What's the reason for moving this file?

File: backend/src/apiserver/client_manager/client_manager.go
Comment (chensun) [reply backend/src/apiserver/client_manager/client_manager.go:224]: +1, `pgx` feels weird, albeit I saw that's the name of the go package you're using. I think `postgres` would be still better here.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (chensun) [inline backend/src/apiserver/client_manager/client_manager.go:591]: This method seems very fragile. I imagine there's could be multiple scenarios matching keyword "exists".
By the way, why do we need to ignore this now? Was it an issue for MySQL in the past or does only applies to Postgres?

File: backend/src/apiserver/client_manager/client_manager.go
Comment (chensun) [inline backend/src/apiserver/client_manager/client_manager.go:438]: nit: fix error message.

File: backend/test/integration/README.md
Comment (chensun) [inline backend/test/integration/README.md:17]: Is it in your next step to automate the test? We can rely on manual test.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [reply backend/src/apiserver/client_manager/client_manager.go:53]: Sounds good, thanks for confirming! We are still keeping the user-facing flag unchanged: https://github.com/kubeflow/pipelines/blob/master/backend/src/cache/client_manager.go#L112-L115

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:591]: I added this because MySQL simply ignores if the foreign key already exist, but PostgreSQL reports an error. A more robust way I can think of is to use an if/else branch for the two SQLs and handle them separately.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:224]: I used this because we are using the package `pgx` as the PostgreSQL driver, and it is needed in the `sql.Open()` function to create the correct DB instance. See the example [here](https://pkg.go.dev/github.com/jackc/pgx/stdlib). I agree it may be quite confusing to people, but I was weighting this against being consistent with the driver name we are using, so I eventually opted for making it clear in the comments to reduce the confusion. WDYT?

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:59]: I agree that they are pretty much the same fields here, but I think it is possible that there can be different values for the two types of DB, as well as different configs available, so it's safer to have one set of values for each of them. For example, the default host name and port number for MySQL and PostgreSQL are different.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:15]: Good point! I moved it because I added integration test for functions here, but it's impossible to test stuff in "main" from a different package in Golang. Reference: https://stackoverflow.com/questions/53997114/how-to-test-the-main-package-in-golang-from-a-test-package

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:303]: I agree. This part was more of a patch work to make it work for now.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:396]: Yes, thanks for finding this! I should add a default user here.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:405]: Yes, thank you :)

File: backend/src/apiserver/client/sql.go
Comment (Linchin) [reply backend/src/apiserver/client/sql.go:60]: `passfile` sounds like a great way to enhance security. Could you tell me more about it? My guess is this will allow users to pass a file containing the password as part of the deployment config. Maybe this can be added together with other configs users can set.

File: backend/src/apiserver/client/sql.go
Comment (Linchin) [reply backend/src/apiserver/client/sql.go:63]: I wonder if this is necessary, since the host address here is simply for communications inside the cluster. Could you tell me more why this is beneficial? Thank you!

File: backend/test/integration/run_tests_locally.sh
Comment (Linchin) [reply backend/test/integration/run_tests_locally.sh:37]: added

File: backend/test/integration/run_tests_locally.sh
Comment (Linchin) [reply backend/test/integration/run_tests_locally.sh:41]: Thank you for finding this. The command for port forwarding is also slightly different for these two DBs. About 4 lines above. I moved the more repetitive lines outside so the script is more succinct.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:268]: Could you help me understand why it is better to use `const` values here? Thank you.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:405]: done

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:438]: Thank you, removed.

File: backend/test/integration/README.md
Comment (Linchin) [reply backend/test/integration/README.md:17]: Ideally we should, and also include PostgreSQL set of tests in the test infra.

File: backend/src/apiserver/client/sql.go
Comment (zijianjoy) [reply backend/src/apiserver/client/sql.go:24]: Makes sense, I am just curious about this change, thank you for the explanation!

File: backend/src/apiserver/client/sql.go
Comment (zijianjoy) [reply backend/src/apiserver/client/sql.go:60]: That is right! So the `passfile` identify the location path to a password, so you don't need to expose the secret via deployment. https://www.postgresql.org/docs/current/libpq-pgpass.html When passing this parameter to postgresql, the `password` field can be omitted.

We can add this configuration in the coming PR, if this makes sense to you.

File: backend/src/apiserver/client/sql.go
Comment (zijianjoy) [reply backend/src/apiserver/client/sql.go:63]: Sure thing, `hostaddr` is helpful when we are connecting to external database, for example: CloudSQL. Yes we can use `host` which is handy for `Unix-domain communication`, but `hostaddr` allows postgresql client to make `TCP/IP communication` instead.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [reply backend/src/apiserver/client_manager/client_manager.go:224]: Okay I think @Linchin 's point is reasonable, we just need to make sure we document this parameter is being used as  a sql driver flag: https://github.com/golang/go/wiki/SQLDrivers.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [reply backend/src/apiserver/client_manager/client_manager.go:268]: DB dialects should be stored in a centralized place, so it helps reviewing, modifying and extending. For example: If we integrate with a new DB type in the future, I hope that I can view the existing syntax differences easily in one place, instead of trying to catch those adhoc variables during implementation. 

Therefore, my recommendation is to define a const for `MYSQL_TEXT_FORMAT` and `PGX_TEXT_FORMAT` in a centralized file: https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/storage/db.go

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [inline backend/src/apiserver/client_manager/client_manager.go:600]: We can define consts for each error message keyword `already exists` and `database exists` in a centralized place: https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/storage/db.go

File: manifests/kustomize/third-party/postgresql/base/kustomization.yaml
Comment (zijianjoy) [inline manifests/kustomize/third-party/postgresql/base/kustomization.yaml:5]: I shouldn't need a PV resource, it is automatically created by PVC when it is bound to a deployment resource.

File: backend/test/integration/run_tests_locally.sh
Comment (Linchin) [reply backend/test/integration/run_tests_locally.sh]: Thank you! I updated the file.

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:406]: Added

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go:268]: Thank you so much for the thourought explanation! I added the constants in `backend/src/apiserver/client/sql.go` instead, because this is a SQL related dependency that client_manager is already using.

File: manifests/kustomize/third-party/postgresql/base/kustomization.yaml
Comment (Linchin) [reply manifests/kustomize/third-party/postgresql/base/kustomization.yaml:5]: I'm not sure about this. I added the PV file because the PVC had trouble creating due to lack of PV. Maybe I was doing something wrong during my deployment. Could you give me more reference on this? Thank you.

**Comment by Linchin** [comment]
part of #9813

File: backend/src/apiserver/client/sql.go
Comment (zijianjoy) [reply backend/src/apiserver/client/sql.go:54]: This allows scenario like connecting to a managed database service. We can discuss about this in future implementation.

File: manifests/kustomize/third-party/postgresql/base/kustomization.yaml
Comment (zijianjoy) [reply manifests/kustomize/third-party/postgresql/base/kustomization.yaml:5]: You can look at the `Events` for the PVC to learn more about why is PV not created. Maybe the disk driver has something wrong when provisioning a PV. But we shouldn't create PV resource manually in codebase. Refer  to MySQL manifest folder: https://github.com/kubeflow/pipelines/tree/master/manifests/kustomize/third-party/mysql/base

File: backend/src/apiserver/client_manager/client_manager.go
Comment (zijianjoy) [inline backend/src/apiserver/client_manager/client_manager.go:406]: Set password default value as `password` here? reference: https://github.com/kubeflow/pipelines/blob/master/manifests/kustomize/third-party/postgresql/base/pg-secret.yaml#L7

File: backend/test/integration/run_tests_locally.sh
Comment (zijianjoy) [inline backend/test/integration/run_tests_locally.sh:17]: These two commands can be combined `set -ex`

File: backend/src/apiserver/client_manager/client_manager.go
Comment (Linchin) [reply backend/src/apiserver/client_manager/client_manager.go]: Good point! Added the password.

File: backend/test/integration/run_tests_locally.sh
Comment (Linchin) [reply backend/test/integration/run_tests_locally.sh]: Thank you for catching this! Just fixed it.

File: manifests/kustomize/third-party/postgresql/base/kustomization.yaml
Comment (Linchin) [reply manifests/kustomize/third-party/postgresql/base/kustomization.yaml]: You are right about this. I tried a fresh deployment without the PV, and it passed the tests. The previous error was likely caused by something else. I removed the pv declaration.

**Review by zijianjoy — COMMENTED** [review]
/lgtm

Great work on this PR, Lingqing!
Waiting for Chen for a final approval.

**Review by chensun — COMMENTED** [review]
/lgtm

Thanks!

**Review by chensun — APPROVED** [review]
/approve

Thanks!

**Comment by iciclespider** [comment]
I get the following error when the db driver name is configured to "gpx":

client_manager.go:428] sql: unknown driver "pgx" (forgotten import?)

According to this, it appears to be a missing import which will register the needed driver: https://stackoverflow.com/a/52791919

**Pull request merged**
